File: scripts/update-readme.js
// @ts-check

import { readFileSync, writeFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import { loadJSON } from './utils.js';

/** @typedef {import('./types.js').StreakData} StreakData */
/** @typedef {import('./types.js').Milestone} Milestone */

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

/**
 * Update stats section in README
 * @param {string} readme - Current README content
 * @param {StreakData} streak - Streak data
 * @returns {string} Updated README content
 */
function updateStats(readme, streak) {
  const stats = `\`\`\`
🔥 Current Streak: ${streak.current_streak} days
🏃 Total Jogging Days: ${streak.total_days}
📅 Last Jog: ${streak.last_updated}
\`\`\``;

  return readme.replace(
    /```\n🔥 Current Streak:[\s\S]*?```/,
    stats
  );
}

/**
 * Update milestones section
 * @param {string} readme - Current README content
 * @param {number} currentStreak - Current streak count
 * @returns {string} Updated README content
 */
function updateMilestones(readme, currentStreak) {
  /** @type {Milestone[]} */
  const milestones = [
    { days: 7, text: "7 days 🌱" },
    { days: 30, text: "30 days 🌿" },
    { days: 100, text: "100 days 🌳" },
    { days: 365, text: "365 days 🏆" }
  ];

  const milestoneText = milestones
    .map(m => currentStreak >= m.days ? `- [x] ${m.text}` : `- [ ] ${m.text}`)
    .join('\n') + '\n';

  return readme.replace(
    /## 🏆 Milestones\n\n[\s\S]*?(?=\n---)/,
    `## 🏆 Milestones\n\n${milestoneText}`
  );
}

/**
 * Main function to update README
 */
function main() {
  // Define file paths
  const streakPath = join(__dirname, '..', 'streak.json');
  const readmePath = join(__dirname, '..', 'README.md');

  // Load data
  /** @type {StreakData} */
  const streak = loadJSON(streakPath, {
    current_streak: 0,
    total_days: 0,
    last_updated: 'Never'
  });

  // Read README
  let readme = readFileSync(readmePath, 'utf8');

  // Update sections
  readme = updateStats(readme, streak);
  readme = updateMilestones(readme, streak.current_streak);

  // Save updated README
  writeFileSync(readmePath, readme);

  console.log('✅ README updated');
}

// Run if called directly
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  main();
}

export { main as updateReadme };

File: scripts/update_readme.py
#!/usr/bin/env python3
import json
import re
from datetime import datetime

def load_json(filename):
    try:
        with open(filename, 'r') as f:
            return json.load(f)
    except:
        return {}

def update_readme():
    # Load all data
    exercises_data = load_json('exercises.json')
    streak_data = load_json('streak.json')
    viz_data = load_json('visualizations.json')
    
    exercises = exercises_data.get('exercises', [])
    
    # Read current README
    with open('README.md', 'r') as f:
        readme = f.read()
    
    # Update stats section
    stats_section = f"""```
🔥 Current Streak: {streak_data.get('current_streak', 0)} days
💪 Total Exercise Days: {streak_data.get('total_days', 0)}
📅 Last Exercise: {streak_data.get('last_updated', 'Never')}
🎯 Goal: 365 days of movement!
```"""
    
    readme = re.sub(r'```\n🔥 Current Streak:.*?```', stats_section, readme, flags=re.DOTALL)
    
    # Update exercise log (last 10 entries)
    recent_exercises = sorted(exercises, key=lambda x: x['date'], reverse=True)[:10]
    
    log_rows = []
    for exercise in recent_exercises:
        log_rows.append(f"| {exercise['date']} | {exercise['activity']} | {exercise['duration']} | {exercise.get('notes', '')} |")
    
    log_section = "\n".join(log_rows) if log_rows else "| No exercises logged yet |"
    
    # Update the exercise log table
    pattern = r'(\| Date \| Activity \| Duration \| Notes \|\n\|------\|----------|----------|-------|\n).*?(\n<!-- DO NOT EDIT ABOVE THIS LINE -->)'
    replacement = f"\\1{log_section}\\2"
    readme = re.sub(pattern, replacement, readme, flags=re.DOTALL)
    
    # Update heatmap if available
    if viz_data.get('heatmap'):
        heatmap_section = f"```\n{viz_data['heatmap']}\n```"
        readme = re.sub(r'### Monthly Heatmap\n<!-- Heatmap will be auto-generated -->\n```\n.*?```', 
                       f"### Monthly Heatmap\n<!-- Heatmap will be auto-generated -->\n{heatmap_section}", 
                       readme, flags=re.DOTALL)
    
    # Update milestones
    current_streak = streak_data.get('current_streak', 0)
    milestones = [
        (7, "7 days streak 🌱"),
        (30, "30 days streak 🌿"),
        (100, "100 days streak 🌳"),
        (365, "365 days streak 🏆")
    ]
    
    milestone_lines = []
    for days, text in milestones:
        if current_streak >= days:
            milestone_lines.append(f"- [x] {text}")
        else:
            milestone_lines.append(f"- [ ] {text}")
    
    milestone_section = "\n".join(milestone_lines)
    readme = re.sub(r'## 🎖️ Milestones\n\n.*?(?=\n##)', 
                   f"## 🎖️ Milestones\n\n{milestone_section}\n", 
                   readme, flags=re.DOTALL)
    
    # Write updated README
    with open('README.md', 'w') as f:
        f.write(readme)
    
    print("✅ Updated README.md")

if __name__ == "__main__":
    update_readme()